src/vs/workbench/browser/parts/editor/multiEditorTabsControl.ts

```diff
@@ -889,29 +890,27 @@ export class MultiEditorTabsControl extends EditorTabsControl {
 			if (editor) {
 				if (e.shiftKey) {
 					let anchor;
-					if (this.lastSelectedEditor && this.groupView.isSelected(this.lastSelectedEditor)) {
+					if (this.lastSingleSelectSelectedEditor && this.tabsModel.isSelected(this.lastSingleSelectSelectedEditor)) {
 						// The last selected editor is the anchor
-						anchor = this.lastSelectedEditor;
+						anchor = this.lastSingleSelectSelectedEditor;
 					} else {
 						// The active editor is the anchor
-						this.lastSelectedEditor = this.groupView.activeEditor!;
+						this.lastSingleSelectSelectedEditor = this.groupView.activeEditor!;
 						anchor = this.groupView.activeEditor!;
 					}
 					this.selectEditorsBetween(editor, anchor);
 				} else if ((e.ctrlKey && !isMacintosh) || (e.metaKey && isMacintosh)) {
-					if (this.groupView.isSelected(editor)) {
-						this.groupView.unSelectEditor(editor);
+					if (this.tabsModel.isSelected(editor)) {
+						this.groupView.unselectEditors([editor]);
 					} else {
-						this.groupView.selectEditor(editor, true);
-						this.lastSelectedEditor = editor;
+						this.groupView.selectEditors([editor], editor);
+						this.lastSingleSelectSelectedEditor = editor;
 					}
 				} else {
 					// Even if focus is preserved make sure to activate the group.
-					this.groupView.openEditor(editor, { preserveFocus, activation: EditorActivation.ACTIVATE }, { selected: this.groupView.isSelected(editor) /* Ensures drag and drop does not remove selection */ });
+					this.groupView.openEditor(editor, { preserveFocus, activation: EditorActivation.ACTIVATE }, { addToSelection: this.tabsModel.isSelected(editor) /* Ensures drag and drop does not remove selection */ });
```

Comment:
I am curious here: if I have multiple tabs selected and I click into a selected tab, I would have expected that because of `addToSelection: this.tabsModel.isSelected(editor)`, this would preserve all of the selection, but it seems the clicked tab is the only selected one. Is our logic flawed here?

Reply:
I'll add a comment here. The reason we use `addToSelection` here is because of drag and drop. This function runs on key_down, so we do not want to unselect the tabs at this moment as the user might be doing a drag and drop operation with multiple tabs. There is a function which runs on key_up which is responsible of removing the selection when there is no drag and drop operation.